### ISSUE TITLE
Improve parallelism settings

### ISSUE THREAD

**Issue by rebeccafair** [pace-neutrons/Euphonic#128]
Currently by default `calculate_qpoint_phonon_modes` has `use_c=False` and `n_threads=1`. This means that by default it uses serial Python (even if the C extension is installed) which is undesirable, and means that to make the most of Euphonic's performance `use_c` and `n_threads` have to be passed all the way through the call stack. This is a particular problem for command line scripts, which are supposed to be quick and easy to use, having to pass these arguments all the time would be cumbersome.

To make this more user friendly, a few changes are proposed:
* `use_c` should be retained as it is good to give users full control if they want it, and it is also useful for testing, but with some changes.
  * `use_c=True` should force use of the C extension, and raise an exception if it isn't installed
  * `use_c` should have a new default `None` which acts as 'auto' and will use the C extension if it is installed, and quietly fall back to Python if it isn't installed
* `n_threads` should also be retained as this fine control could be useful for e.g. running multiple Euphonic instances at once, but with a new default `None`. If `n_threads=None`, the environment variable `EUPHONIC_NUM_THREADS` will be read, and if the environment variable isn't set, `n_threads` will be set to <s>1</s> the output of `multiprocessing.cpu_count()`. This allows users to either set the environment variable once, then not have to worry about the number of threads, or get a sensible default. This is particularly useful for command line tools.

**Comment by ajjackson** [comment]
This all looks good to me.

Might be worth discussing whether the fallback default should indeed be 1; another option would be
```python
import multiprocessing
default_n_threads = multiprocessing.cpu_count()
```
On my i7 that returns 8, so it doesn't distinguish between true cores and hyper-threading. But I get better performance on `calculate_qpoint_phonon_modes` with hyperthreading so that's not a bad thing!

**Comment by rebeccafair** [comment]
That's a good point, I do worry that occasionally `cpu_count()` will give a count that is too high and decrease performance, but I haven't actually seen this much in practice, some quick profiling on my laptop (`cpu_count() = 4`) and IDAaaS (`cpu_count() = 30`) shows there is still improvement. There are probably far more cases where it provides a benefit, so I'll use `cpu_count()` as the default.

**Comment by ajjackson** [comment]
If we do introduce some serial chunking to limit memory usage during DOS/powder calculations, then I suggest we target 1GB/thread or less as workstations generally have a bit over 1GB/thread of memory.

**Comment by rebeccafair** [comment]
Another thing I've noticed, is that Numpy's maths libraries do their own multithreading by default, which we don't really want when Euphonic is doing its own, better scaling, multithreading over q-points. It's not always clear which environment variable to set as it depends on what Numpy is linked to, but for example if I have 

|n_threads|Time(s) OMP_NUM_THREADS unset|Time(s) OMP_NUM_THREADS=1 |
| --------- | --------------------------------------- | ----------------------------------- |
|1| | |
|1| | |
|1| | |
|1| | |
|1| | |
|1| | |
|1| | |
|1| | |

**Comment by ajjackson** [comment]
Interesting! Well that is a compelling reason not to assume that OMP_NUM_THREADS  is the preferred number of threads for Euphonic; in fact the user can optimise a combination of EUPHONIC_NUM_THREADS and OMP_NUM_THREADS.

Some interesting discussion here, machine-learning people also getting frustrated with numpy parallelism. https://github.com/numpy/numpy/issues/11826 It looks like a package has been developed to help with this problem: https://pypi.org/project/threadpoolctl/

**Comment by rebeccafair** [comment]
Nice find! I wasn't aware of that library. Its very easy to implement, and doesn't have any other dependencies so I might just add it straight away, as I can't think of any obvious downsides? On IDAaaS it shows _almost_ as good performance as setting OMP_NUM_THREADS manually, but not quite as good, I'm not sure why:

|n_threads|Time(s) OMP_NUM_THREADS unset|Time(s) OMP_NUM_THREADS=1 |Time(s) OMP_NUM_THREADS unset, with threadpool_limits|
| --------- | --------------------------------------- | ----------------------------------- | ----------------------------------- |
|1| 13.948| 9.030|9.105|
|2| 7.624| 4.710|4.790|
|4| 4.687| 2.572|2.632|
|8| 3.314| 1.543|1.783|
|12| 2.835| 1.391|1.647|
|16| 2.614 | 1.428|1.730|
|20| 2.445| 1.468|1.779|
|24| 2.318| 1.528|1.846|
|28| 2.226| 1.600|1.905|
|30| 2.213| 1.629|1.933|

I just need to check it plays nice with before committing to it.

**Comment by ajjackson** [comment]
Maybe it takes 0.1s to operate?

**Comment by rebeccafair** [comment]
That's what I thought, but tried with 50,000 q-points to see and got this:

|n_threads|Time(s) OMP_NUM_THREADS unset|Time(s) OMP_NUM_THREADS=1 |Time(s) OMP_NUM_THREADS unset, with threadpool_limits|
| --------- | --------------------------------------- | ----------------------------------- | ----------------------------------- |
|1| 69.583 | 44.861|45.365|
|2| 36.927 | 23.670|24.081|
|4| 22.058 | 12.913|13.372|
|8| 16.381 | 7.904|8.965|
|12| 14.058 | 6.941|8.154|
|16| 12.656 | 7.077|8.426|
|20|  11.696 | 7.364|8.704|
|24| 10.935 | 7.688|9.076|
|28| 10.693 | 7.922|9.339|
|30| 10.425 | 8.098|9.394|

So I'm not really sure why... it's better than not including it at all though

**Comment by ajjackson** [comment]
Yup, I'll take that gain!

**Comment by rebeccafair** [comment]
Closed by #129
